mitiq/cdr/test_clifford_training_data.py

```diff
@@ -0,0 +1,254 @@
+# Copyright (C) 2021 Unitary Fund
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Tests for generating (near) Clifford circuits."""
+import pytest
+import numpy as np
+
+import cirq
+from cirq.circuits import Circuit
+
+from mitiq.cdr.clifford_training_data import (
+    _is_clifford_angle,
+    _is_clifford,
+    is_clifford,
+    _map_to_near_clifford,
+    _select,
+    _replace,
+    _closest_clifford,
+    _random_clifford,
+    _angle_to_probabilities,
+    _probabilistic_angle_to_clifford,
+    count_non_cliffords,
+    generate_training_circuits,
+    _CLIFFORD_ANGLES,
+)
+
+
+def random_x_z_cnot_circuit(qubits, n_moments, random_state) -> Circuit:
+    angles = np.linspace(0.0, 2 * np.pi, 8)
+    oneq_gates = [
+        gate(a) for a in angles for gate in (cirq.ops.rx, cirq.ops.rz)
+    ]
+    gate_domain = {oneq_gate: 1 for oneq_gate in oneq_gates}
+    gate_domain.update({cirq.ops.CNOT: 2})
+
+    return cirq.testing.random_circuit(
+        qubits=qubits,
+        n_moments=n_moments,
+        op_density=1.0,
+        gate_domain=gate_domain,
+        random_state=random_state,
+    )
+
+
+def test_generate_training_circuits():
+    circuit = random_x_z_cnot_circuit(
+        cirq.LineQubit.range(3), n_moments=5, random_state=1
+    )
+    assert not is_clifford(circuit)
+
+    (clifford_circuit,) = generate_training_circuits(
+        circuit, num_training_circuits=1, fraction_non_clifford=0.0
+    )
+    assert is_clifford(clifford_circuit)
+
+
+@pytest.mark.parametrize("method", ("uniform", "gaussian"))
+def test_select_all(method):
+    q = cirq.LineQubit(0)
+    ops = [cirq.ops.rz(0.01).on(q), cirq.ops.rz(-0.77).on(q)]
+    indices = _select(
+        ops, 0.0, method=method, random_state=np.random.RandomState(1)
+    )
+    assert np.allclose(indices, np.array(list(range(len(ops)))))
+
+
+@pytest.mark.parametrize("method", ("uniform", "gaussian"))
+def test_select_some(method):
+    n = 10  # Number to select.
+    q = cirq.GridQubit(1, 1)
+    ops = [cirq.ops.rz(a).on(q) for a in np.random.randn(n)]
```

Comment:
```suggestion
    q = cirq.GridQubit(1, 1)
    rnd_state = np.random.RandomState(1)
    ops = [cirq.ops.rz(a).on(q) for a in rnd_state.randn(n)]
```
If we have bad luck this test could fail. In this way it becomes deterministic.